### PR TITLE
Datepicker mapstatetoprops optin

### PR DESCRIPTION
**Pull request by Ledoux**
given #24 

@alexsegura @Atala I used DatePicker Component to illustrate a bit how you can optimize
a Component refresh workflow tx to redux and the lifecycle of react

### CODE REVIEW

**Comment by Ledoux** [comment]
first thing to read is about the mapStateToProps modifications
```js
export default compose(
  withRouter,

  // 1. mapStateToProps is called for anykind of triggered actions.
  // so doing moment computations in the mapStateToProps time
  // is a cost but not useful, if actually the last action
  // was dealing with completely something else related to DatePicker

  // 2. keeping this idea, it is always better in mapStateToProps to return
  // values that are directly variables we can shallowCompare
  // if they differ from the previous redux state

  // 3. In that manner, the component shouldUpdates/renders just only
  // for the values of the redux state that concerns this Component
  connect(({ deliveryDate, restaurant: { availabilities } }) =>
    ({ availabilities, deliveryDate }),
  { setDeliveryDate })
)(DatePicker)
```

**Comment by Ledoux** [comment]
then look at the WillMount and WillReceiveProps hooks

```js
// 1. when shouldComponentUpdate said that ok,
  // the Component has to refresh given the new redux state
  // we can do the computation at WillMount and WillReceiveProps time
  // 2. we keep in the local
  // state of the component these computed date and time variables
  handleSetDateAndTime ({ availabilities, deliveryDate }) {
    let date, time
    if (!deliveryDate) {
      const first = availabilities[0]
      const firstMoment = moment(first)
      date = firstMoment.format('YYYY-MM-DD')
      time = firstMoment.format('HH:mm')
    } else {
      const deliveryDateMoment = moment(deliveryDate)
      date = deliveryDateMoment.format('YYYY-MM-DD')
      time = deliveryDateMoment.format('HH:mm')
    }
    this.setState({ date, time })
  }
  componentWillMount () {
    this.handleSetDateAndTime(this.props)
  }
  componentWillReceiveProps (nextProps) {
    // no need to compute if actually the component updates
    // for restaurant.availabilities changes
    if (nextProps.deliveryDate !== this.props.deliveryDate) {
      this.handleSetDateAndTime(nextProps)
    }
  }
```

**Comment by Ledoux** [comment]
and finally at the DidMount and DidUpdate times

```js
  // 1. The Component still needs to trigger itself a new action
  // given its new local state, which is possible to control
  // at DidMount and DidUpdate time
  // 2. So once date and time are stored in the local state
  // we trigger the redux action setDeliveryDate, but we make sure
  // to call it once
  handleSetDeliveryDate () {
    const { date, time } = this.state
    this.props.setDeliveryDate(date + ' ' + time + ':00')
  }
  componentDidMount () {
    this.handleSetDeliveryDate()
  }
  componentDidUpdate (nextState) {
    // only trigger the action once
    // when date and/or time have changed in the local state
    if (nextState.date !== this.state.date || nextState.time !== this.state.time) {
      this.handleSetDeliveryDate()
    }
  }
```

File: src/components/DatePicker.js [hunk 1]
```diff
@@ -1,94 +1,157 @@
...
 
   render() {
 
-    const { availabilities } = this.props;
-    const days = _.groupBy(availabilities, date => moment(date).format('YYYY-MM-DD'));
-    const dates = _.keys(days);
+    const { availabilities, date, time } = this.props
+    const { availableTimes } = this.state
+
+    const days = _.groupBy(availabilities, date =>
```
Comment (Atala) [inline src/components/DatePicker.js:103]: Should by `groupBy` not `_.groupBy` :p

**Comment by Atala** [comment]
Thanks for the clear explanation! 

Merging into master.

**Pull request merged**